src/openstack_cloud/openstack_manager.py

```diff
@@ -1509,20 +1518,71 @@ def _issue_reconciliation_metric(
         except IssueMetricEventError:
             logger.exception("Failed to issue Reconciliation metric")
 
-    def flush(self) -> int:
+    def flush(self, mode: FlushMode = FlushMode.FLUSH_IDLE) -> int:
         """Flush Openstack servers.
 
+        1. Kill the processes depending on flush mode.
+        2. Get unhealthy runners after process purging.
+        3. Delete unhealthy runners.
+
+        Args:
+            mode: The mode to determine which runner to flush.
+
         Returns:
             The number of runners flushed.
         """
         logger.info("Flushing OpenStack all runners")
         with _create_connection(self._cloud_config) as conn:
+            self._kill_runner_processes(conn=conn, mode=mode)
             runner_by_health = self._get_openstack_runner_status(conn)
             remove_token = self._github.get_runner_remove_token(path=self._config.path)
-            runners_to_delete = (*runner_by_health.healthy, *runner_by_health.unhealthy)
             self._remove_runners(
                 conn=conn,
-                instance_names=runners_to_delete,
+                instance_names=runner_by_health.unhealthy,
                 remove_token=remove_token,
             )
-            return len(runners_to_delete)
+        return len(runner_by_health.unhealthy)
+
+    def _kill_runner_processes(self, conn: OpenstackConnection, mode: FlushMode) -> None:
```

Comment:
Is this needed?
Would the runner remove command that takes the remove token do this already?

Reply:
Would you be referring to the `self._remove_runners` command?

The reason for having this is because from the moment where we perform health-checks via ssh until we actually delete the runners have quite a bit of a time lag in which the jobs can be picked up by the runners (since the deletion of runners are sequential and if OpenStack decides to take time to perform the action, the bigger the discrepancy from the truth).

This function reduces the racy implementation of health-check -> deletion by determining on spot whether the runner process should be stopped (thereby no longer receiving jobs), and then cleaning up via health check afterwards.